docs/performance/inserts/sequences.rst

```diff
@@ -0,0 +1,179 @@
+
+*************
+ Using UUIDs
+*************
+
+This option involves declaring a column as follows:
+
+.. code:: psql
+
+   TEXT DEFAULT gen_random_text_uuid() PRIMARY KEY
+
+**Pros:** Globally unique, no risk of conflicts if merging things from different
+tables/environments
+
+**Cons:** No order guarantee, not as human-friendly as numbers
+
+************************
+ Use UUIDv7 identifiers
+************************
+
+`Version 7 UUIDs`_ are a relatively new kind of UUIDs which feature a
+time-ordered value. We can use these in CrateDB with an UDF_ with the code from
+https://github.com/nalgeon/uuidv7/pull/37 (update once merged)
+
+**Pros:** Same as gen_random_text_uuid above but almost sequential
+
+**Cons:** not as human-friendly as numbers and slight performance impact from
+UDF use
+
+*********************************
+ Use IDs from an external system
+*********************************
+
+In cases where data is flowing to CrateDB from another system, we can declare
+the column on the CrateDB side without any defaults and just ingest the
+auto-incrementing values coming from the source database.
+
+See
+https://cratedb.com/blog/replicating-data-from-other-databases-to-cratedb-with-debezium-and-kafka
+for an example.
+
+*********************
+ Implement sequences
+*********************
+
+This approach involves a table to keep the latest values that have been consumed
+and client side code to keep it up-to-date in a way that guarantees unique
+values even when many ingestion processes run in parallel.
+
+**Pros:** Can have any arbitrary type of sequences, (we may for instance want to
+increment values by 10 instead of 1 - prefix values with a year number - combine
+numbers and letters - etc)
+
+**Cons:** Need logic for the optimistic update implemented
+client-side, the sequences table becomes a bottleneck so not suitable for
+high-velocity ingestion scenarios
+
+We will first create a table to keep the latest values for our sequences:
+
+.. code:: psql
+
+   CREATE TABLE sequences (
+           name TEXT PRIMARY KEY,
+           last_value BIGINT
+   ) CLUSTERED INTO 1 SHARDS;
+
+We will then initialize it with one new sequence at 0:
+
+.. code:: psql
+
+   INSERT INTO sequences (name,last_value)
+   VALUES ('mysequence',0);
+
+And we are going to do an example with a new table defined as follows:
+
+.. code:: psql
+
+   CREATE TABLE mytable (
+           id BIGINT PRIMARY KEY,
+           field1 TEXT
+   );
+
+The Python code below reads the last value used from the sequences table, and
+then attempts an `optimistic UPDATE`_ with a ``RETURNING`` clause, if a contending
+process already consumed the identity nothing will be returned so our process
+will retry until a value is returned, then it uses that value as the new ID for
+the record we are inserting into the ``mytable`` table.
+
+.. code:: python
+
+   import records
+
+   db = records.Database("crate://")
+   sequence_name = "mysequence"
+
+   while True:
```

Comment:
That reads dangerous, potentially implementing a busy loop, but I guess you know what you are doing?
Still, shall we possibly add a timeout on this identifier-polling operation, so the loop may not accidentally spin forever?